src/api/mod.rs

```diff
@@ -1,5 +1,21 @@
 //! API specific documentation.
 
+#[cfg(feature = "spaces")]
+// FIXME: Async methods aren't supported in traits. We could
+// use a normal method that returns Pin<Box<_>> monstrosity,
+// but is that okay? How can we work around this?
+macro_rules! impl_execute {
```

Comment:
I'm still working on this btw. I know it's not that great to duplicate code like this. Maybe I'd go for something else once I implement object-related thingies, but nothing comes to mind as of now.

Reply:
Could `impl Trait` help us with the return value?

Reply:
The problem is that all builder structs may need to have their own `execute` method (right now, we have `Executable` trait which is pretty neat, but since all the `execute` methods are async, we can't use them in traits unless we box stuff).

Reply:
There's nothing inherently wrong with boxing things.  I know there are some strategies around pinning as well.

Reply:
I know, but there's the guilt of making an user's decision for them. :laughing: So, you okay with the trait way? We'll have a `Box<Pin<Foo>>` or something in an `AsyncExecutable` trait. We could later merge it with `Executable`.